hasktorch/src/Torch/Typed/Device.hs

```diff
@@ -0,0 +1,14 @@
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
+module Torch.Typed.Device where
+
+import           GHC.TypeLits
+
+data DeviceType = CPU | CUDA
```

Comment:
One thing we should consider is that at some point in the future PyTorch will support registration of new devices at run-time, and so having a closed ADT might not be a good solution. I don't know how to do it differently though, since you cannot have kind-classes (type-level type-classes).

Reply:
hmm, so I thought of that. And I think that defining all possible device types in line 9 is fine and necessary. same with device indices. I'm using `Nat` for that, and hence the number of devices is not restricted at compile time.
that means that functions like `toCUDA` will have to return an option of a moved tensor.

Reply:
oh, wait, I think I misunderstood you.
are you saying that future torch versions will support adding device runtimes at runtime?
in that case, defining all possible device types in line 9 would be impossible.

Reply:
Yes, that's what I meant. I don't think hot-swapping GPUs is that important, but being open to new backends is (you're e.g. already missing TPU).